core/arch/arm/pta/benchmark.c

```diff
@@ -0,0 +1,199 @@
+			p[0].memref.buffer,
+			virt_to_phys(p[0].memref.buffer));
+	bench_ts_global = p[0].memref.buffer;
+
+	/* send back to the optee linux kernel module */
+	res = rpc_reg_global_buf(OPTEE_MSG_RPC_CMD_BENCH_REG_NEW,
+			virt_to_phys((void *)bench_ts_global),
+			sizeof(struct tee_ts_global) +
+			sizeof(struct tee_ts_cpu_buf) *
+			bench_ts_global->cores);
+
+	mutex_unlock(&bench_reg_mu);
+
+	return res;
+}
+
+static TEE_Result get_benchmark_memref(uint32_t type,
+				TEE_Param p[TEE_NUM_PARAMS])
+{
+	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_VALUE_OUTPUT) ||
+		(TEE_PARAM_TYPE_GET(type, 1) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 2) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 3) != TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	mutex_lock(&bench_reg_mu);
+
+	DMSG("Sending back timestamp buffer paddr = %p\n",
+		(void *)virt_to_phys(bench_ts_global));
+	p[0].value.a = virt_to_phys(bench_ts_global);
+
+	mutex_unlock(&bench_reg_mu);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result unregister_benchmark(uint32_t type,
+				TEE_Param p[TEE_NUM_PARAMS] __unused)
+{
+	TEE_Result res;
+
+	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 1) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 2) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 3) != TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	mutex_lock(&bench_reg_mu);
+
+	DMSG("Unregistering benchmark, timestamp buffer paddr = %p\n",
+		(void *)virt_to_phys(bench_ts_global));
+	bench_ts_global = NULL;
+
+	res = rpc_reg_global_buf(OPTEE_MSG_RPC_CMD_BENCH_REG_DEL, 0, 0);
+
+	mutex_unlock(&bench_reg_mu);
+
+	return res;
+}
+
+static TEE_Result invoke_command(void *pSessionContext __unused,
+		uint32_t nCommandID, uint32_t nParamTypes,
+		TEE_Param pParams[TEE_NUM_PARAMS])
+{
+	switch (nCommandID) {
+	case BENCHMARK_CMD_REGISTER_MEMREF:
+		return register_benchmark_memref(nParamTypes, pParams);
+	case BENCHMARK_CMD_GET_MEMREF:
+		return get_benchmark_memref(nParamTypes, pParams);
+	case BENCHMARK_CMD_UNREGISTER:
+		return unregister_benchmark(nParamTypes, pParams);
+	default:
+		break;
+	}
+
+	return TEE_ERROR_BAD_PARAMETERS;
+}
+
+pseudo_ta_register(.uuid = BENCHMARK_UUID, .name = TA_NAME,
+		   .flags = PTA_DEFAULT_FLAGS,
+		   .invoke_command_entry_point = invoke_command);
+
+void bm_timestamp(void)
+{
+	struct tee_ts_cpu_buf *cpu_buf;
+	struct tee_time_st ts_data;
+	uint64_t ts_i;
+	void *ret_addr;
+	uint32_t cur_cpu;
+	uint32_t exceptions;
+
+	if (!bench_ts_global)
+		return;
+
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	cur_cpu = get_core_pos();
+
+	/* fill timestamp data */
+	if (cur_cpu >= bench_ts_global->cores) {
+		thread_unmask_exceptions(exceptions);
+		return;
+	}
+
+	ret_addr = __builtin_return_address(0);
+
+	cpu_buf = &bench_ts_global->cpu_buf[cur_cpu];
+	ts_i = __sync_fetch_and_add(&cpu_buf->head, 1);
```

Comment:
> Foreign interruptions (actually all interrupts) are already masked, so you can be pretty sure you can atomically update head value with even a complex SW sequence.

Agree. This came from the version, when there was only common timestamp buffer for all CPUs. I'll leave `__sync_fetch_and_add` only for NW EL0 apps (optee_client).

> Also, if head is lower than tail an reaches tail when incremented, it should update tail and maybe raise an overrun flag.

Right, we have already discussed it with Joakim, and he suggested to leave it for now as it is, and add support of overrun detection in further PRs. As for me, it's better just to add additioanl field to `struct tee_time_st` with will contain timestamp id (actually, it can be the `head` value, as it's incrementing all the time), so we can detect an over-run when parsing results by the actual timestamp analyzing app, which will detect a bit gap in timestamp IDs.